### PR TITLE
Move constructor helpers to EAMixins

### PR DESCRIPTION
**Pull request by jbrockmendel**
Takes the place of #21843, porting constructor helpers.  While this is in review I'm going to start porting tests in earnest.

### CODE REVIEW

File: pandas/core/arrays/period.py [hunk 1]
```diff
@@ -8,7 +8,7 @@
 from pandas._libs.tslib import NaT, iNaT
 from pandas._libs.tslibs.period import (
     Period, IncompatibleFrequency, DIFFERENT_FREQ_INDEX,
-    get_period_field_arr, period_asfreq_arr)
+    get_period_field_arr, period_asfreq_arr, _quarter_to_myear)
```
Comment (jreback) [inline pandas/core/arrays/period.py]: minor nit on using privatized cython functions here (later PR)

File: pandas/core/arrays/period.py [hunk 2]
```diff
@@ -157,6 +160,25 @@ def _from_ordinals(cls, values, freq=None):
         result._freq = Period._maybe_convert_freq(freq)
         return result
 
+    @classmethod
+    def _generate_range(cls, start, end, periods, freq, fields):
```
Comment (jreback) [inline pandas/core/arrays/period.py]: future pls add doc-string

File: pandas/core/arrays/period.py [hunk 3]
```diff
@@ -371,3 +393,102 @@ def _add_comparison_methods(cls):
 
 
 PeriodArrayMixin._add_comparison_methods()
+
+
+# -------------------------------------------------------------------
+# Constructor Helpers
+
+def _get_ordinal_range(start, end, periods, freq, mult=1):
+    if com._count_not_none(start, end, periods) != 2:
```
Comment (jreback) [inline pandas/core/arrays/period.py]: future doc-string & comments

File: pandas/core/arrays/period.py [hunk 4]
```diff
@@ -371,3 +393,102 @@ def _add_comparison_methods(cls):
...
+            data = np.arange(start.ordinal, start.ordinal + periods, mult,
+                             dtype=np.int64)
+    else:
+        data = np.arange(start.ordinal, end.ordinal + 1, mult, dtype=np.int64)
+
+    return data, freq
+
+
+def _range_from_fields(year=None, month=None, quarter=None, day=None,
+                       hour=None, minute=None, second=None, freq=None):
```
Comment (jreback) [inline pandas/core/arrays/period.py]: future doc-string & comments

File: pandas/core/arrays/period.py [hunk 5]
```diff
@@ -371,3 +393,102 @@ def _add_comparison_methods(cls):
...
+        base, mult = frequencies.get_freq_code(freq)
+        arrays = _make_field_arrays(year, month, day, hour, minute, second)
+        for y, mth, d, h, mn, s in compat.zip(*arrays):
+            ordinals.append(libperiod.period_ordinal(
+                y, mth, d, h, mn, s, 0, 0, base))
+
+    return np.array(ordinals, dtype=np.int64), freq
+
+
+def _make_field_arrays(*fields):
```
Comment (jreback) [inline pandas/core/arrays/period.py]: doc-string

File: pandas/core/arrays/timedelta.py [hunk 6]
```diff
@@ -83,6 +101,52 @@ def __new__(cls, values, freq=None):
 
         return result
 
+    @classmethod
+    def _generate(cls, start, end, periods, freq, closed=None, **kwargs):
```
Comment (jreback) [inline pandas/core/arrays/timedelta.py]: we call this ``_generate_range`` above, maybe in future can share *some* code in base class

File: pandas/core/arrays/timedelta.py [hunk 7]
```diff
@@ -198,3 +301,55 @@ def to_pytimedelta(self):
...
+        if not hasnans:
+            result = result.astype('int64')
+        return result
+
+
+# ---------------------------------------------------------------------
+# Constructor Helpers
+
+def _generate_regular_range(start, end, periods, offset):
+    stride = offset.nanos
```
Comment (jreback) [inline pandas/core/arrays/timedelta.py]: future doc-string

**Review by jreback — APPROVED** [review]
lgtm. the only thing is rename _generate -> _generate_range, everything else can be future things

**Comment by jbrockmendel** [comment]
> rename _generate -> _generate_range, everything else can be future things

Will do.

**Comment by jbrockmendel** [comment]
Thoughts on renaming (not this PR) arrays.period --> arrays.periods and arrays.timedelta --> arrays.timedeltas?  For the core.indexes files datetimes and timedeltas are plural, period is not.

**Comment by jbrockmendel** [comment]
> rename _generate -> _generate_range, everything else can be future things

This caused some failures, reverted it until the next pass.

**Comment by jreback** [comment]
thanks !

**Pull request merged**